i6_models/config.py

```diff
@@ -51,3 +54,28 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class ModuleFactoryV1(Generic[ConfigType, ModuleType]):
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Type[ModuleType]
+    cfg: ConfigType
+
+    def __call__(self) -> ModuleType:
+        """Constructs an instance of the given module class"""
+        return self.module_class(self.cfg)
+
+    def __post_init__(self) -> None:
+        parameters = signature(self.module_class).parameters.values()
+        assert len(parameters) == 1
+        cfg_parameter = next(iter(parameters))
+        typeguard.check_type(self.cfg, cfg_parameter.annotation)
```

Comment:
Can you give an example where this "fails"?

Reply:
I think in general using `signature` is not totally reliable and it is mostly intended to be used for inspection and debugging.

My examples were just examples. I can imagine there are more cases.

I just don't think such code is really necessary. I would keep it simple. But if you really want to have it, then keep it.

However, I think better than such a check would be to fix the typing, as I said before. Then PyCharm and other tools could also print you warnings.

Reply:
I don't know. I haven't really tried.

I just asked the [question on StackOverflow](https://stackoverflow.com/questions/76572323/combine-protocol-with-base-class-requirement-type-intersection). If we don't get a good answer there, the [official `typing` package on GitHub](https://github.com/python/typing) might be another good place to ask, or the [official `mypy` package on GitHub](https://github.com/python/mypy).

Reply:
Yes we decided that we want this check so that any error is detected early on. The check has been made a bit more general now so that all (or at least most) use-cases should be covered. If this breaks for some specific module class later we can adjust the check.